lib/kennel/models/record.rb

```diff
@@ -129,25 +135,53 @@ def build_json
         }.compact
       end
 
+      # One of three outcomes:
+      # - throws (on non-validation error)
+      # - returns, validation_errors is not empty, and as_json is nil (invalid)
+      # - returns, validation_errors is empty, and as_json is not nil (valid)
+      def prepare
+        @validation_errors = []
+        json = nil
+
+        begin
+          json = build_json
+          (id = json.delete(:id)) && json[:id] = id
+          validate_json(json) if validate
+        rescue StandardError
```

Comment:
why do we need to rescue here ? if we let it raise we get a nice backtrace too ?

Reply:
The thinking here is that because `invalid!` used to raise, no further processing happened beyond that. But that's not true any more; `invalid` sets state, and returns, potentially to allow more validation code to run. Since we've already encountered an error, we enter a kind of "all bets are off" state. For example, one might have:

```
def validate_json(data)
  data.key?(:tags) or invalid! "Need tags"

  data[:tags].all? { ... } or invalid! "..."
end

i.e. as soon as we've hit one validation error, we shouldn't be surprised if we crash beyond that.

This is what the rescue is for: to swallow errors _if_ they happen after we've already seen a validation error.

Reply:
Re. backtrace: at the moment on this branch, the output looks a bit like this:

```
Finding applicable projects and parts ... 2.11s
Preparing json ... 0.03s

user-segment-service:14a91074 type 'metric alert' is deprecated, please set to a different type (e.g. 'query alert')
user-segment-service:14a91074 priority needs to be an Integer
user-segment-service:user-segment-service-k8s-pods critical and value used in query must match

rake aborted!
Kennel::GenerationAbortedError: Kennel::GenerationAbortedError
<big useless backtrace>
```

i.e. each error is a single line, with the tracking id, and no backtrace. Do you want something different?

Reply:
yeah, above seems best ... I want backtraces for debugging, but maybe that can be a later feature